### PR TITLE
[libc++] Adjust some of the [rand.dist] critical values that are too strict

### PR DESCRIPTION
**Pull request by MattStephanson**
Adjust some of the [rand.dist] critical values that are too strict

 - Most critical values are determined empirically by running each test 51
   times with a different PRNG seed and finding the smallest symmetric interval
   around the median that contains 90% of the sample means, variances, etc.

 - For the Kolmogorov-Smirnov tests, the alpha=0.1 critical value for large N
   is 1.224/sqrt(N).

 - For normally distributed variates, the sample kurtosis is distributed as
   Normal(0, 24/N). For N=1e5, this gives a 90% confidence interval of
   0+/-0.0255. For Binomial(40, 0.25), which is approximately normal, the
   kurtosis is -0.0167, so the relative 90% CI is large, on the order of
   0.0255/0.0167 = 153%. In most cases the distribution of the sample kurtosis
   isn't known analytically, but similarly large relative tolerances can be
   expected if the kurtosis is near zero.

### CODE REVIEW

File: libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp [hunk 1]
```diff
@@ -151,8 +151,8 @@ void test3() {
     double x_kurtosis = (1-6*d.p()*(1-d.p())) / x_var;
     assert(std::abs((mean - x_mean) / x_mean) < 0.01);
     assert(std::abs((var - x_var) / x_var) < 0.01);
-    assert(std::abs((skew - x_skew) / x_skew) < 0.03);
-    assert(std::abs((kurtosis - x_kurtosis) / x_kurtosis) < 0.3);
```
Comment (mordante) [inline libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp:155]: Why did you change this test?

File: libcxx/test/std/numerics/rand/rand.dist/rand.dist.norm/rand.dist.norm.lognormal/eval.pass.cpp [hunk 2]
```diff
@@ -70,8 +70,8 @@ test1()
                       3*std::exp(2*sqr(d.s())) - 6;
     assert(std::abs((mean - x_mean) / x_mean) < 0.01);
     assert(std::abs((var - x_var) / x_var) < 0.01);
-    assert(std::abs((skew - x_skew) / x_skew) < 0.05);
-    assert(std::abs((kurtosis - x_kurtosis) / x_kurtosis) < 0.25);
+    assert(std::abs((skew - x_skew) / x_skew) < 0.1);
+    assert(std::abs((kurtosis - x_kurtosis) / x_kurtosis) < 1.9);
```
Comment (mordante) [inline libcxx/test/std/numerics/rand/rand.dist/rand.dist.norm/rand.dist.norm.lognormal/eval.pass.cpp:74]: This change looks rather big, did you happen to do any investigation why this changed so much?

**Review by mordante — COMMENTED** [review]
Thanks for working on this. Can you update the commit message how you determined the new values; I guess empirical, but it would be good to know for certain.

I noticed the same when I worked on [P0952R2](https://wg21.link/P0952R2). (That patch is still WIP and I got lower on my priority list.)

File: libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp (hunk 1 above)
Comment (MattStephanson) [reply libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp:155]: Oops, I was debating changing it to an absolute test because the expected kurtosis is close to zero (-0.017). 2.0 should be the relative bound.

For reference, if you approximate the binomial distribution as normal, you would expect the sample kurtosis to have a standard deviation of $\sqrt{24/N} \approx 0.0155$. That gives a relative 90% confidence interval of $\sqrt{2}\mathrm{erf}^{-1}(0.9)\cdot 0.0155 / 0.017 \approx 1.5$, so I don't think the empirically observed value of 2.0 is unreasonable.

File: libcxx/test/std/numerics/rand/rand.dist/rand.dist.norm/rand.dist.norm.lognormal/eval.pass.cpp (hunk 2 above)
Comment (MattStephanson) [reply libcxx/test/std/numerics/rand/rand.dist/rand.dist.norm/rand.dist.norm.lognormal/eval.pass.cpp:74]: The expected kurtosis is 0.0039, so I think that explains the loose relative tolerance.

These variates aren't normal, but just as a rough landmark, if they were, the standard deviation of the sample kurtosis would be 0.00490, so I don't think a 190% tolerance is unreasonable.

**Comment by MattStephanson** [comment]
> Is in not better to have smaller values on this?

Not exactly. The statistics of a sample (mean, variance, etc.) are themselves random variables with some distribution. So except in degenerate cases, there's always some chance that a correct generator will give an out-of-tolerance result. There's also a tradeoff because if you want a smaller tolerance, you need more samples to justify rejecting the null hypothesis that the sample comes from the desired distribution. Even these larger tolerances that I'm proposing only approximate a 90% confidence interval, so if the underlying U[0,1) engine was truly random, you would still expect these tests to fail 10% of the time _even if the distribution generator was completely correct_.

File: libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp (hunk 1 above)
Comment (mordante) [reply libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp:155]: Can you add something along these lines in the commit message. That give future readers more information why these changes are valid.

**Review by mordante — APPROVED** [review]
> Is in not better to have smaller values on this?

I kind of agree, however these are statistical tests and there is no motivation why these tolerances were picked. For example 45a999719b45e1ef00d928c54ca25ef8ef6889ff

I noticed the tolerances not to work when implementing a new paper.
 
> If so, we may want to keep some testing to ensure our implementation doesn't slip? But otherwise, this change seems reasonable.

I think we could add a test where we test some specific expected values. To me this is outside the scope of this patch; unless @MattStephanson likes to do it.

I'm happy with this patch after addressing my last comment. Please give @EricWF a few days to respond before landing this patch.

File: libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp (hunk 1 above)
Comment (mordante) [reply libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp:155]: @MattStephanson can you update the commit message? Then the patch is good to land. Thanks.

Comment (MattStephanson) [reply libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp:155]: @mordante I did update it the last time I force-pushed [788fc37](https://github.com/llvm/llvm-project/commit/788fc376ccd4084d4e7ba916da3ddae5a7944c46). If it's still missing something, could you please be more specific? Thanks!

Comment (MattStephanson) [reply libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp:155]: Oh, okay. I think I have it now?

Comment (mordante) [reply libcxx/test/std/numerics/rand/rand.dist/rand.dist.bern/rand.dist.bern.bin/eval.pass.cpp:155]: Yes great! Thanks. I'll land the patch.

**Comment by mordante** [comment]
I noticed your email is hidden, could you make that public?

**Comment by MattStephanson** [comment]
> I noticed your email is hidden, could you make that public?

Okay.
